File: sections/68_tree_decompositions/problems/c_marked_pair_distances/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using int64 = long long;
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int n, q;
  if (!(cin >> n >> q)) return 0;
  vector<vector<pair<int, int>>> g(n);
  for (int i = 1, u, v, w; i < n; i++) {
    cin >> u >> v >> w;
    --u;
    --v;
    g[u].push_back({v, w});
    g[v].push_back({u, w});
  }
  int log = 1;
  while ((1 << log) <= n) log++;
  vector<vector<int>> up(log, vector<int>(n));
  vector<int> tin(n), tout(n), depth(n);
  vector<int64> distance(n);
  int timer = 0;
  vector<tuple<int, int, int, int>> stack{{0, 0, 0, 0}};
  while (!stack.empty()) {
    auto [u, p, w, state] = stack.back();
    stack.pop_back();
    if (!state) {
      tin[u] = timer++;
      up[0][u] = p;
      for (int j = 1; j < log; j++) up[j][u] = up[j - 1][up[j - 1][u]];
      stack.push_back({u, p, w, 1});
      for (auto it = g[u].rbegin(); it != g[u].rend(); ++it)
        if (it->first != p) {
          depth[it->first] = depth[u] + 1;
          distance[it->first] = distance[u] + it->second;
          stack.push_back({it->first, u, it->second, 0});
        }
    } else
      tout[u] = timer - 1;
  }
  auto ancestor = [&](int u, int v) { return tin[u] <= tin[v] && tout[v] <= tout[u]; };
  auto lca = [&](int u, int v) {
    if (ancestor(u, v)) return u;
    if (ancestor(v, u)) return v;
    for (int j = log - 1; j >= 0; j--)
      if (!ancestor(up[j][u], v)) u = up[j][u];
    return up[0][u];
  };
  vector<char> marked(n);
  vector<int> count(n), virtual_parent(n);
  while (q--) {
    int k;
    cin >> k;
    vector<int> nodes(k);
    for (int& u : nodes) {
      cin >> u;
      --u;
      marked[u] = 1;
    }
    sort(nodes.begin(), nodes.end(), [&](int a, int b) { return tin[a] < tin[b]; });
    int original = nodes.size();
    for (int i = 1; i < original; i++) nodes.push_back(lca(nodes[i - 1], nodes[i]));
    sort(nodes.begin(), nodes.end(), [&](int a, int b) { return tin[a] < tin[b]; });
    nodes.erase(unique(nodes.begin(), nodes.end()), nodes.end());
    vector<int> st;
    for (int u : nodes) {
      while (!st.empty() && !ancestor(st.back(), u)) st.pop_back();
      virtual_parent[u] = st.empty() ? -1 : st.back();
      st.push_back(u);
      count[u] = marked[u];
    }
    int64 answer = 0;
    for (int i = nodes.size() - 1; i >= 0; i--) {
      int u = nodes[i];
      int p = virtual_parent[u];
      if (p != -1) {
        answer += (distance[u] - distance[p]) * count[u] * (k - count[u]);
        count[p] += count[u];
      }
    }
    cout << answer << '\n';
    for (int u : nodes) marked[u] = 0;
  }
}

File: sections/68_tree_decompositions/problems/c_marked_pair_distances/solution.py
import sys


def main():
    tokens = sys.stdin.buffer.read().split()
    if not tokens:
        return
    idx = 0
    n = int(tokens[idx])
    q = int(tokens[idx + 1])
    idx += 2
    graph = [[] for _ in range(n)]
    for _ in range(n - 1):
        u = int(tokens[idx]) - 1
        v = int(tokens[idx + 1]) - 1
        w = int(tokens[idx + 2])
        idx += 3
        graph[u].append((v, w))
        graph[v].append((u, w))
    log = n.bit_length()
    up = [[0] * n for _ in range(log)]
    tin = [0] * n
    tout = [0] * n
    distance = [0] * n
    timer = 0
    stack = [(0, 0, 0)]
    while stack:
        u, p, state = stack.pop()
        if state == 0:
            tin[u] = timer
            timer += 1
            up[0][u] = p
            for j in range(1, log):
                up[j][u] = up[j - 1][up[j - 1][u]]
            stack.append((u, p, 1))
            for v, w in reversed(graph[u]):
                if v != p:
                    distance[v] = distance[u] + w
                    stack.append((v, u, 0))
        else:
            tout[u] = timer - 1

    def ancestor(u, v):
        return tin[u] <= tin[v] <= tout[u]

    def lca(u, v):
        if ancestor(u, v):
            return u
        if ancestor(v, u):
            return v
        for j in range(log - 1, -1, -1):
            if not ancestor(up[j][u], v):
                u = up[j][u]
        return up[0][u]

    marked = [False] * n
    count = [0] * n
    parent = [-1] * n
    out = []
    for _ in range(q):
        k = int(tokens[idx])
        idx += 1
        nodes = [int(x) - 1 for x in tokens[idx : idx + k]]
        idx += k
        for u in nodes:
            marked[u] = True
        nodes.sort(key=tin.__getitem__)
        nodes += [lca(nodes[i - 1], nodes[i]) for i in range(1, k)]
        nodes = sorted(set(nodes), key=tin.__getitem__)
        st = []
        for u in nodes:
            while st and not ancestor(st[-1], u):
                st.pop()
            parent[u] = st[-1] if st else -1
            st.append(u)
            count[u] = int(marked[u])
        answer = 0
        for u in reversed(nodes):
            p = parent[u]
            if p != -1:
                answer += (distance[u] - distance[p]) * count[u] * (k - count[u])
                count[p] += count[u]
        out.append(str(answer))
        for u in nodes:
            marked[u] = False
    print("\n".join(out))


if __name__ == "__main__":
    main()
